src/serializers/mod.rs

```diff
@@ -120,15 +121,18 @@ impl SchemaSerializer {
             false,
             fallback,
         );
-        let v = self.serializer.to_python(value, include, exclude, &extra)?;
+        let v = match duck_typed_serialization {
+            false => self.serializer.to_python(value, include, exclude, &extra)?,
+            true => infer::infer_to_python(value, include, exclude, &extra)?,
```

Comment:
Does this work for nested models beyond the first level? I would think we'd want to add this onto the `extra` or similar so that it can be respected in the serializer for deeper-nested items.

Reply:
Example demonstrating that this doesn't work on nested types:
```python
from pydantic import BaseModel


class Inner(BaseModel):
    x: int


class SubInner(Inner):
    y: int


class Outer(BaseModel):
    inner: Inner


m = Outer(inner=SubInner(x=1, y=2))
print(m.__pydantic_serializer__.to_python(m, duck_typed_serialization=True))
# Current:
#> {'inner': {'x': 1}}

# Desired:
#> {'inner': {'x': 1, 'y': 2}}
```
If I understand correctly, I think this would need to be refactored to add an attribute to `extra` indicating whether to do duck typed serialization, to enure it gets passed all the way down

Reply:
Here's the test I came up with:

```python


def test_duck_typed_serialization() -> None:
    class Parent:
        x: int
        nested: Union['Parent', None]

        __pydantic_core_schema__: core_schema.CoreSchema
        __pydantic_validator__: SchemaValidator
        __pydantic_serializer__: SchemaSerializer

    Parent.__pydantic_core_schema__ = core_schema.model_schema(
        Parent,
        core_schema.model_fields_schema(
            {
                'x': core_schema.model_field(core_schema.int_schema()),
                'nested': core_schema.model_field(
                    core_schema.nullable_schema(core_schema.definition_reference_schema('parent'))
                ),
            }
        ),
        ref='parent',
    )
    Parent.__pydantic_validator__ = SchemaValidator(Parent.__pydantic_core_schema__)
    Parent.__pydantic_serializer__ = SchemaSerializer(Parent.__pydantic_core_schema__)

    class Child(Parent):
        y: str = 'maybe a secret'

    Child.__pydantic_core_schema__ = core_schema.definitions_schema(
        core_schema.model_schema(
            Child,
            core_schema.model_fields_schema(
                {
                    'x': core_schema.model_field(core_schema.int_schema()),
                    'nested': core_schema.model_field(
                        core_schema.nullable_schema(core_schema.definition_reference_schema('parent'))
                    ),
                    'y': core_schema.model_field(core_schema.str_schema()),
                }
            ),
        ),
        [Parent.__pydantic_core_schema__],
    )
    Child.__pydantic_validator__ = SchemaValidator(Child.__pydantic_core_schema__)
    Child.__pydantic_serializer__ = SchemaSerializer(Child.__pydantic_core_schema__)

    s = SchemaSerializer(core_schema.list_schema(Parent.__pydantic_core_schema__))

    child = Child.__pydantic_validator__.validate_python({'x': 1, 'nested': None, 'y': 'hopefully not a secret'})

    res = s.to_python(
        [
            Child.__pydantic_validator__.validate_python({'x': 1, 'nested': child, 'y': 'hopefully not a secret'}),
            Parent.__pydantic_validator__.validate_python({'x': 2, 'nested': child}),
        ],
        duck_typed_serialization=True,
    )

    assert res == [
        {'x': 1, 'nested': {'x': 1, 'nested': None, 'y': 'hopefully not a secret'}, 'y': 'hopefully not a secret'},
        {'x': 2, 'nested': {'x': 1, 'nested': None, 'y': 'hopefully not a secret'}},
    ]

    res = s.to_python(
        [
            Child.__pydantic_validator__.validate_python({'x': 1, 'nested': child, 'y': 'hopefully not a secret'}),
            Parent.__pydantic_validator__.validate_python({'x': 2, 'nested': child}),
        ],
        duck_typed_serialization=False,
    )

    assert res == [
        {'x': 1, 'nested': {'x': 1, 'nested': None}},
        {'x': 2, 'nested': {'x': 1, 'nested': None}},
    ]
```

Unfortunately making this work is not going to be easy. Not only do we need to pass that information down, any time we encounter an "inner" validator we would have to ask it how to serialize (e.g. check if it has a `__pydantic_serializer__`). I think that would have a performance hit even for the cases where we are not doing that, to the point where it would probably make more sense to build two `SchemaSerializers` upfront: one that tries to duck type serialize every object and another that doesn't.